xc/device.ml

```diff
@@ -607,13 +607,39 @@ let qemu_media_change ~xs device _type params =
 
 	if is_upstream_qemu device.frontend.domid
 	then begin
+		let open Qmp in
 		let cd = "ide1-cd1" in
-		let qmp_cmd =
-			if params = ""
-			then (Qmp.Command(None, Qmp.Eject (cd, Some true)))
-			else (Qmp.Command(None, Qmp.Change (cd, params, None))) in
-		qmp_write device.frontend.domid qmp_cmd
-	end
+		match params with
+		| "" ->
+			let qmp_cmd = Command(None, Eject (cd, Some true)) in
+			qmp_write device.frontend.domid qmp_cmd
+		| _ ->
+			let c = Qmp_protocol.connect (Printf.sprintf "/var/run/xen/qmp-libxl-%d" device.frontend.domid) in
```

Comment:
I expect this pattern of "let c = Qmp_procotol.connect(domid) finally (fun ()-> ..my_function..) (fun ()-> Qmp_protocol.close c)" to be repeated many times in xenopsd. It would be very useful if we could transform this into a function "let with_qmp" with parameters (my_function c) so that this whole pattern can be easily rewritten as "with_qmp (fun c ->  my_function c )". Another advantage of using with_qmp is that it can be used to control concurrent access to the qmp protocol: as far as I know, only one qmp connection can be opened at any time, so we could easily add a Mutex inside with_qmp to make sure that at most one qmp connection is open at any time.